src/aiida_quantumespresso/workflows/protocols/pw/base.yaml

```diff
@@ -25,7 +25,7 @@ default_inputs:
                 smearing: cold
                 degauss: 0.01
             ELECTRONS:
-                electron_maxstep: 80
+                electron_maxstep: 50
```

Comment:
Shall we set it a bit larger, e.g. `60`, as to include 99% of cases instead? Or you say that eventually one would just restart the calculation if the slope is good?

Reply:
Then again... I was redoing the analysis on the slopes, but then instead of looking at the slopes over _all_ steps, look only at the 50 we actually calculated in case we'd set `electron_maxstep` to 50:

![Screenshot 2023-11-27 at 22 41 38](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/b4a4bf86-523b-4203-9e1b-a195fe4e93ba)

Now we see quite a few with slopes > -0.1 that _do_ converge. Looking at the evolution of some of their scf accuracies:

![image](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/e39377b6-725a-475f-9842-cf5b555c0cfe)

We can see that indeed, only after around step 50 the SCF "catches its groove", and the convergence begins accelerating. Maybe this would also happen if we restart from the charge density and reduce the mixing, but that isn't certain and it'd most likely be slower.

You can play round with the "max nstep" to consider. Of course as you increase it, you have fewer cases where the slope before "max nstep" is > -0.1 and the calculation converges, but you have fewer cases of successful convergence between "max step" and 80 steps as well. 

So we could ask two questions: if I would have set `electron_maxstep` to X, and used -0.1 as the slope threshold

* what's the chance that we did an "incorrect stop", i.e. the slope is > -0.1, but the calculation _converges_ in 80 steps?
* what's the chance that we did an "incorrect restart", i.e. the slope is < -0.1, but the calculation _fails to converge_ in 80 steps?

| max nstep | # incorrect stop | % incorrect stop | # incorrect restart | % incorrect restart |
|-|-|-|-|-|
| 70 | 13 | 6.3 | 1153 | 46.2 |
| 60 | 27 | 5.4 | 1292 | 51.7 |
| 50 | 54 | 5.3 | 1425 | 57.0 |
| 40 | 61 | 2.9 | 1584 | 63.4 |
| 30 | 72 | 1.6 | 1700 | 68.1 |
|20 | 80 | 0.7 | 1919 | 76.8 |

But frankly, these % data are also skewed by the fact that we only have data up to step 80, i.e. if we had data up to 200 steps the % incorrect stop would go down for all "max nstep", but most likely more for 70 than for 40.

So to do this analysis "right", we'd need to gather statistics for a larger `electron_maxstep`. Moreover, the question is what happens if we "incorrectly stop" a calculation and decide to e.g. reduce the mixing. It's all quite a lot of analysis without too much benefit in the end, I think. 60 seems like a reasonable choice to me, so I'll pick that unless there are objections.

One note though: it's clear from the "scf accuracy" curves above that extrapolation may not always give a reasonable value. I'd keep it simple: if the slope < -0.1, just double the `electron_maxstep` to 120 and restart from the last charge density.

Reply:
Thanks @mbercx for the analysis. I must say I couldn't completely follow the arguments. But the figure you showed is also interesting. Maybe we should compute the slope only on e.g. 20 of the last steps, and look at the slope, and decide whether to restart or to change the mixing mode?

Reply:
Yeah, it was a little late, so maybe my explanation wasn't the best. ^^ Let me summarize:

In [my previous analysis](https://github.com/aiidateam/aiida-quantumespresso/issues/961#issuecomment-1825585559) I used the slope over all 80 steps to see if we can decide after 50 steps to continue with more steps or switch to another strategy. That's of course not correct, since at step 50 we'd only have information about the first 50. If we look at the slope distributions for the first 50 steps (first image in my post above), we can see that there are more cases where selecting -0.1 as the threshold would lead to our approach switching to another strategy (e.g. adapting `mixing_beta`), or doing an "incorrect stop".

I was curious if you could "optimize" the choice for `electron_maxstep`, i.e. minimize the % of incorrect stops. But to judge this properly I think we need to have data for larger `electron_maxstep` runs, since we don't really know which of the failed runs _would_ have converged if we had run with more than 80 steps. Also: we don't know if switching to e.g. a lower `mixing_beta` would improve or worsen the convergence after step 50. So perhaps we shouldn't overdo this analysis for now.

> Maybe we should compute the slope only on e.g. 20 of the last steps

I was also thinking of this approach. Here is the analysis:

```
$ ./gather_stats.py slopes --last 20 --max-nstep 50 workchain/PBEsol/scf
Report: Number of completed pw.x with > 50 SCF steps: 3524
Report: Calculating slopes from step 30 to 50.
Report: Incorrect stops: 130 | 12.7% of 1026 successful runs with > 50 steps
Report: Incorrect restarts: 834 | 33.4% of 2498 failed runs
```

![Screenshot 2023-11-28 at 17 26 04](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/1ae99c17-9ef6-43ac-807b-7c0e445a5f37)

So based on this setup (information about 80 steps, trial "max nstep" 50, slope threshold -0.1), only checking the last 20 steps increases the # of incorrect stops, but reduces the # of incorrect restarts. Below are some examples of runs where we would incorrectly stop:

![image](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/00354867-fbcf-4c86-8cd8-df03e4ba1886)

So not sure if this really improves things. I suppose it depends what we prioritise: convergence or saving resources.

Reply:
Thanks @mbercx ! Ok, I see now. Well, the the point is also whether a change of strategy (e.g. mixing beta) will ever produce significant change in the first 50 (or 60, or whatever) iterations. There is a high chance to keep trying changing strategy, while having just some iterations more would solve the convergence. 
So I don't know how much we are going to save resources if we are not really sure to fix with a certain probability the problem. So I would vote to even _increase_ such maximum number of iterations (which btw on QE I think the default is to 100).